File: filesystem/deno/src/core/fs-object.ts
import { Directory } from './directory.ts';
import { FsObjectType } from './fs-object-type.enum.ts';

export abstract class FsObject {
	constructor(
		public readonly name: string,
		public readonly parent: FsObject | null,
		public readonly fsObjectType: FsObjectType,
	) {}
}

File: filesystem/go/core/fs-object.go
package core

import "errors"

type FsObject interface {
	GetName() string
	GetParent() (FsObject, error)
	GetType() FsObjectType
}

type AbstractFsObject struct {
	Name   string
	Parent *AbstractFsObject
	Type   FsObjectType
}

func (o *AbstractFsObject) GetName() string {
	return o.Name
}

func (o *AbstractFsObject) GetParent() (FsObject, error) {
	if o.Parent == nil {
		return nil, DirectoryHasNoParentError()
	}
	return o.Parent, nil
}

func (o *AbstractFsObject) GetType() FsObjectType {
	return o.Type
}

func DirectoryHasNoParentError() error {
	return errors.New("Directory has no parent")
}
